File: 2023/ocaml/lib/day1.ml
open Core

let reverse str =
  let rec loop idx =
    match idx with
    | 0 -> String.make 1 str.[0]
    | _ -> String.make 1 str.[idx] ^ loop (idx - 1)
  in
  loop (String.length str - 1)
;;

let digit_word str idx =
  let words =
    [ "one"; "two"; "three"; "four"; "five"; "six"; "seven"; "eight"; "nine" ]
  in
  let str = String.sub str ~pos:idx ~len:(String.length str - idx) in
  let opt_word i word =
    if String.is_substring_at str ~pos:0 ~substring:word then Some (i + 1) else None
  in
  let matches = List.filter_map ~f:Fun.id (List.mapi ~f:opt_word words) in
  match matches with
  | [] -> None
  | item :: _rest -> Some item
;;

let first_digit_or_word str =
  let rec loop idx =
    if idx < String.length str
    then (
      let c = str.[idx] in
      match c with
      | '0' .. '9' -> Some (int_of_string (String.make 1 c))
      | _ ->
        (match digit_word str idx with
         | Some s -> Some s
         | None -> loop (idx + 1)))
    else None
  in
  loop 0
;;

let last_digit_or_word str =
  let rec loop idx =
    if idx >= 0
    then (
      let c = str.[idx] in
      match c with
      | '0' .. '9' -> Some (int_of_string (String.make 1 c))
      | _ ->
        (match digit_word str idx with
         | Some s -> Some s
         | None -> loop (idx - 1)))
    else None
  in
  loop (String.length str - 1)
;;

let first_digit str =
  let rec loop idx =
    if idx < String.length str
    then (
      let c = str.[idx] in
      match c with
      | '0' .. '9' -> Some (int_of_string (String.make 1 c))
      | _ -> loop (idx + 1))
    else None
  in
  loop 0
;;

let first_and_last_digit str =
  let first = first_digit str in
  let second = first_digit (reverse str) in
  match first, second with
  | Some first, Some second -> Some ((first * 10) + second)
  | _ -> None
;;

let first_and_last_digit_or_word str =
  let first = first_digit_or_word str in
  let second = last_digit_or_word str in
  match first, second with
  | Some first, Some second -> Some ((first * 10) + second)
  | _ -> None
;;

open Utils

let part1 lines =
  let arr = List.map ~f:first_and_last_digit lines in
  let sum = sum_list (remove_nones arr) in
  Printf.sprintf "%d\n" sum
;;

let part2 lines =
  let arr = List.map ~f:first_and_last_digit_or_word lines in
  let sum = sum_list (remove_nones arr) in
  Printf.sprintf "%d\n" sum
;;

File: 2023/rust/src/day1.rs
use std::{iter::Rev, ops::Range};

use crate::{Day, TestInput};

fn first_digit(line: &[u8], mut indices: impl Iterator<Item = usize>) -> Option<i32> {
    indices.find_map(|s| {
        line[s]
            .is_ascii_digit()
            .then(|| (line[s] as char).to_digit(10).unwrap() as i32)
    })
}

fn first_digit_or_word(line: &[u8], mut indices: impl Iterator<Item = usize>) -> Option<i32> {
    const WORD_MAP: [(&str, i32); 18] = [
        ("one", 1),
        ("1", 1),
        ("two", 2),
        ("2", 2),
        ("three", 3),
        ("3", 3),
        ("four", 4),
        ("4", 4),
        ("five", 5),
        ("5", 5),
        ("six", 6),
        ("6", 6),
        ("seven", 7),
        ("7", 7),
        ("eight", 8),
        ("8", 8),
        ("nine", 9),
        ("9", 9),
    ];

    indices.find_map(|s| {
        WORD_MAP
            .iter()
            .find_map(|(word, value)| line[s..].starts_with(word.as_bytes()).then_some(*value))
    })
}

fn find_digits<F, B>(lines: &[String], forward: F, backward: B) -> i32
where
    F: Fn(&[u8], Range<usize>) -> Option<i32>,
    B: Fn(&[u8], Rev<Range<usize>>) -> Option<i32>,
{
    lines
        .iter()
        .map(String::as_bytes)
        .map(|l| {
            let range = 0..l.len();
            let first = forward(l, range.clone());
            let last = backward(l, range.rev());

            first.unwrap() * 10 + last.unwrap()
        })
        .sum()
}

pub struct Day1;
impl Day for Day1 {
    fn date() -> (i32, i32) {
        (2023, 1)
    }

    fn part1(lines: &[String]) -> String {
        format!("{}", find_digits(lines, first_digit, first_digit))
    }

    fn part2(lines: &[String]) -> String {
        format!(
            "{}",
            find_digits(lines, first_digit_or_word, first_digit_or_word)
        )
    }

    fn test_inputs() -> (TestInput, TestInput) {
        return (
            TestInput {
                input: "
                    1abc2
                    pqr3stu8vwx
                    a1b2c3d4e5f
                    treb7uchet
                ",
                result: "142",
            },
            TestInput {
                input: "
                    two1nine
                    eightwothree
                    abcone2threexyz
                    xtwone3four
                    4nineeightseven2
                    zoneight234
                    7pqrstsixteen
                ",
                result: "281",
            },
        );
    }
}
